pkg/signup/service/signup_service.go

```diff
@@ -131,6 +146,53 @@ func (s *ServiceImpl) newUserSignup(ctx *gin.Context) (*toolchainv1alpha1.UserSi
 	return userSignup, nil
 }
 
+/*
+IsPhoneVerificationRequired determines whether phone verification is required
+
+Returns true in the following cases:
+1. Captcha configuration is disabled
+2. The captcha token is invalid
+3. Captcha failed with an error or the assessment failed
+
+Returns false in the following cases:
+1. Overall verification configuration is disabled
+2. Captcha is enabled and the assessment is successful
+*/
+func IsPhoneVerificationRequired(captchaChecker captcha.Assessor, ctx *gin.Context) (bool, float32) {
+	cfg := configuration.GetRegistrationServiceConfig()
+	if !cfg.Verification().Enabled() {
+		return false, -1
+	}
+
+	if !cfg.Verification().CaptchaEnabled() {
+		return true, -1
+	}
+
+	if ctx == nil || ctx.Request == nil {
+		log.Error(ctx, nil, "no request in context")
```

Comment:
This should never happen, right? I mean if the context is nil or there is no request then it's a bug in our code. If so then I would return an error and fail rather then log and ignore.

Reply:
I added this for unit tests actually, the code fails before this if the context is nil. So I think we can leave it as it is or just remove it and adjust to unit tests to always set ctx to a non-nil value?

Reply:
What do you mean it fails before that? Where does it fail? I see you return "true" meaning the phone verification required and nothing fails. Or I'm missing something?

Reply:
What I meant is `IsPhoneVerificationRequired` is called here https://github.com/rajivnathan/registration-service/blob/94d46dd9ae1de2fd4839b3c8512618045d4ff6ed/pkg/signup/service/signup_service.go#L97 but even before it is called there's lots of places where we try to get things from `ctx` so it would fail much earlier https://github.com/rajivnathan/registration-service/blob/recaptcha/pkg/signup/service/signup_service.go#L70-L81. Basically, I don't think we really need to worry about ctx being nil but the ctx.Request nil check is still fine to have.

Reply:
ok